File: oracle-benchmark-tool/oracle_benchmark.py
#!/usr/bin/env python3
"""
Oracle Performance Benchmark Tool

Compares read performance between:
  1) A single wide table with 100 columns
  2) Three normalized tables joined together (same data, split across 3 tables)

Read patterns tested:
  - SELECT specific columns (5 columns)
  - SELECT through a VIEW
  - SELECT * (all columns)

Usage:
    python oracle_benchmark.py --host <host> --port <port> --service <service> --user <user> --password <password>

Requirements:
    pip install oracledb tabulate
"""

import argparse
import os
import random
import string
import statistics
import sys
import time
from contextlib import contextmanager

try:
    import oracledb
except ImportError:
    print("ERROR: 'oracledb' package is required. Install with: pip install oracledb")
    sys.exit(1)

try:
    from tabulate import tabulate
except ImportError:
    tabulate = None

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

NUM_COLUMNS = 100          # total number of data columns
NUM_ROWS = 10_000          # rows to insert
ITERATIONS = 10            # how many times each query is executed for averaging
FETCH_ALL = True           # whether to fetch all rows (True) or just execute (False)

# For the 3-table split we divide the 100 columns roughly equally
SPLIT = (34, 33, 33)       # columns per table (must sum to NUM_COLUMNS)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def random_string(length: int = 20) -> str:
    return "".join(random.choices(string.ascii_letters + string.digits, k=length))


def col_name(index: int) -> str:
    """Return a column name like COL_001, COL_002, ..."""
    return f"COL_{index:03d}"


def get_connection(args):
    """Create an Oracle connection from CLI args or environment variables."""
    dsn = oracledb.makedsn(
        host=args.host or os.getenv("ORACLE_HOST", "localhost"),
        port=int(args.port or os.getenv("ORACLE_PORT", "1521")),
        service_name=args.service or os.getenv("ORACLE_SERVICE", "FREEPDB1"),
    )
    return oracledb.connect(
        user=args.user or os.getenv("ORACLE_USER", "benchmark"),
        password=args.password or os.getenv("ORACLE_PASSWORD", "benchmark"),
        dsn=dsn,
    )


@contextmanager
def timer():
    """Context manager that records elapsed wall-clock time in seconds."""
    result = {"elapsed": 0.0}
    start = time.perf_counter()
    yield result
    result["elapsed"] = time.perf_counter() - start


def drop_if_exists(cursor, obj_type: str, name: str):
    """Drop an Oracle object if it exists (TABLE, VIEW, etc.)."""
    try:
        cursor.execute(f"DROP {obj_type} {name} PURGE" if obj_type == "TABLE"
                       else f"DROP {obj_type} {name}")
    except oracledb.DatabaseError:
        pass


def print_section(title: str):
    print(f"\n{'=' * 70}")
    print(f"  {title}")
    print(f"{'=' * 70}")


def format_results(results: list[dict]):
    """Pretty-print benchmark results."""
    if tabulate:
        headers = results[0].keys()
        rows = [r.values() for r in results]
        print(tabulate(rows, headers=headers, tablefmt="grid", floatfmt=".4f"))
    else:
        for r in results:
            parts = [f"{k}: {v}" for k, v in r.items()]
            print(" | ".join(parts))


# ---------------------------------------------------------------------------
# Benchmark: Single wide table
# ---------------------------------------------------------------------------

SINGLE_TABLE = "BENCH_SINGLE"
SINGLE_VIEW = "BENCH_SINGLE_V"


def create_single_table(cursor):
    """Create a single table with 100 VARCHAR2 columns + an ID."""
    drop_if_exists(cursor, "VIEW", SINGLE_VIEW)
    drop_if_exists(cursor, "TABLE", SINGLE_TABLE)

    cols = ", ".join(f"{col_name(i)} VARCHAR2(50)" for i in range(1, NUM_COLUMNS + 1))
    ddl = f"CREATE TABLE {SINGLE_TABLE} (ID NUMBER PRIMARY KEY, {cols})"
    cursor.execute(ddl)

    # Create a view that selects 5 specific columns
    view_cols = ", ".join(col_name(i) for i in range(1, 6))
    cursor.execute(
        f"CREATE VIEW {SINGLE_VIEW} AS SELECT ID, {view_cols} FROM {SINGLE_TABLE}"
    )
    print(f"  Created table  {SINGLE_TABLE} with {NUM_COLUMNS} columns")
    print(f"  Created view   {SINGLE_VIEW}")


def insert_single_table(conn, cursor):
    """Insert NUM_ROWS rows of random data into the single table."""
    cols = ", ".join(col_name(i) for i in range(1, NUM_COLUMNS + 1))
    placeholders = ", ".join(f":{i}" for i in range(1, NUM_COLUMNS + 2))  # +1 for ID
    sql = f"INSERT INTO {SINGLE_TABLE} (ID, {cols}) VALUES ({placeholders})"

    batch_size = 1000
    for start in range(0, NUM_ROWS, batch_size):
        rows = []
        for row_id in range(start + 1, min(start + batch_size, NUM_ROWS) + 1):
            row = [row_id] + [random_string(20) for _ in range(NUM_COLUMNS)]
            rows.append(row)
        cursor.executemany(sql, rows)
    conn.commit()
    print(f"  Inserted {NUM_ROWS} rows into {SINGLE_TABLE}")


def benchmark_single_table(cursor) -> list[dict]:
    """Run 3 read patterns on the single wide table and return timing results."""
    select_cols = ", ".join(col_name(i) for i in range(1, 6))
    queries = {
        "SELECT (5 cols)": f"SELECT ID, {select_cols} FROM {SINGLE_TABLE}",
        "VIEW (5 cols)":   f"SELECT * FROM {SINGLE_VIEW}",
        "SELECT *":        f"SELECT * FROM {SINGLE_TABLE}",
    }

    results = []
    for label, sql in queries.items():
        timings = []
        for _ in range(ITERATIONS):
            with timer() as t:
                cursor.execute(sql)
                if FETCH_ALL:
                    cursor.fetchall()
            timings.append(t["elapsed"])

        results.append({
            "Pattern": label,
            "Min (s)": min(timings),
            "Max (s)": max(timings),
            "Avg (s)": statistics.mean(timings),
            "Median (s)": statistics.median(timings),
            "Stdev (s)": statistics.stdev(timings) if len(timings) > 1 else 0.0,
        })
    return results


# ---------------------------------------------------------------------------
# Benchmark: Three normalized tables with JOINs
# ---------------------------------------------------------------------------

MULTI_TABLES = ("BENCH_MULTI_A", "BENCH_MULTI_B", "BENCH_MULTI_C")
MULTI_VIEW = "BENCH_MULTI_V"


def _col_ranges():
    """Return (start, end) column index ranges for each of the 3 tables."""
    s1 = 1
    e1 = SPLIT[0]
    s2 = e1 + 1
    e2 = e1 + SPLIT[1]
    s3 = e2 + 1
    e3 = e2 + SPLIT[2]
    return (s1, e1), (s2, e2), (s3, e3)


def create_multi_tables(cursor):
    """Create 3 tables that together hold the same 100 columns, linked by ID."""
    drop_if_exists(cursor, "VIEW", MULTI_VIEW)
    for tbl in MULTI_TABLES:
        drop_if_exists(cursor, "TABLE", tbl)

    ranges = _col_ranges()
    for tbl, (start, end) in zip(MULTI_TABLES, ranges):
        cols = ", ".join(f"{col_name(i)} VARCHAR2(50)" for i in range(start, end + 1))
        ddl = f"CREATE TABLE {tbl} (ID NUMBER PRIMARY KEY, {cols})"
        cursor.execute(ddl)
        print(f"  Created table {tbl} with {end - start + 1} columns (COL_{start:03d}..COL_{end:03d})")

    # Create a view that joins all 3 tables on ID and picks 5 columns
    view_cols = ", ".join(f"a.{col_name(i)}" for i in range(1, 6))
    cursor.execute(f"""
        CREATE VIEW {MULTI_VIEW} AS
        SELECT a.ID, {view_cols}
        FROM {MULTI_TABLES[0]} a
        JOIN {MULTI_TABLES[1]} b ON a.ID = b.ID
        JOIN {MULTI_TABLES[2]} c ON a.ID = c.ID
    """)
    print(f"  Created view  {MULTI_VIEW} (JOIN of all 3 tables)")


def insert_multi_tables(conn, cursor):
    """Insert the same random data into the 3 normalized tables."""
    ranges = _col_ranges()
    batch_size = 1000

    for start_row in range(0, NUM_ROWS, batch_size):
        # Generate a batch of full rows first
        batch_data = []
        for row_id in range(start_row + 1, min(start_row + batch_size, NUM_ROWS) + 1):
            all_cols = [random_string(20) for _ in range(NUM_COLUMNS)]
            batch_data.append((row_id, all_cols))

        for tbl_idx, (tbl, (cs, ce)) in enumerate(zip(MULTI_TABLES, ranges)):
            num_cols = ce - cs + 1
            cols = ", ".join(col_name(i) for i in range(cs, ce + 1))
            placeholders = ", ".join(f":{j}" for j in range(1, num_cols + 2))
            sql = f"INSERT INTO {tbl} (ID, {cols}) VALUES ({placeholders})"

            rows = []
            for row_id, all_cols in batch_data:
                row = [row_id] + all_cols[cs - 1: ce]
                rows.append(row)
            cursor.executemany(sql, rows)

    conn.commit()
    print(f"  Inserted {NUM_ROWS} rows into each of {', '.join(MULTI_TABLES)}")


def benchmark_multi_tables(cursor) -> list[dict]:
    """Run 3 read patterns using JOINs across the 3 tables and return timing results."""
    select_cols = ", ".join(f"a.{col_name(i)}" for i in range(1, 6))

    all_cols_parts = []
    for tbl_alias, (cs, ce) in zip(("a", "b", "c"), _col_ranges()):
        for i in range(cs, ce + 1):
            all_cols_parts.append(f"{tbl_alias}.{col_name(i)}")
    all_cols = ", ".join(all_cols_parts)

    join_clause = (
        f"FROM {MULTI_TABLES[0]} a "
        f"JOIN {MULTI_TABLES[1]} b ON a.ID = b.ID "
        f"JOIN {MULTI_TABLES[2]} c ON a.ID = c.ID"
    )

    queries = {
        "SELECT (5 cols) + JOIN": f"SELECT a.ID, {select_cols} {join_clause}",
        "VIEW (5 cols) + JOIN":   f"SELECT * FROM {MULTI_VIEW}",
        "SELECT * + JOIN":        f"SELECT a.ID, {all_cols} {join_clause}",
    }

    results = []
    for label, sql in queries.items():
        timings = []
        for _ in range(ITERATIONS):
            with timer() as t:
                cursor.execute(sql)
                if FETCH_ALL:
                    cursor.fetchall()
            timings.append(t["elapsed"])

        results.append({
            "Pattern": label,
            "Min (s)": min(timings),
            "Max (s)": max(timings),
            "Avg (s)": statistics.mean(timings),
            "Median (s)": statistics.median(timings),
            "Stdev (s)": statistics.stdev(timings) if len(timings) > 1 else 0.0,
        })
    return results


# ---------------------------------------------------------------------------
# Cleanup
# ---------------------------------------------------------------------------

def cleanup(cursor):
    """Drop all benchmark objects."""
    drop_if_exists(cursor, "VIEW", SINGLE_VIEW)
    drop_if_exists(cursor, "TABLE", SINGLE_TABLE)
    drop_if_exists(cursor, "VIEW", MULTI_VIEW)
    for tbl in MULTI_TABLES:
        drop_if_exists(cursor, "TABLE", tbl)
    print("  Cleaned up all benchmark objects.")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def parse_args():
    parser = argparse.ArgumentParser(
        description="Oracle read-performance benchmark: single wide table vs 3-table JOIN"
    )
    parser.add_argument("--host", default=None, help="Oracle host (default: env ORACLE_HOST or localhost)")
    parser.add_argument("--port", default=None, help="Oracle port (default: env ORACLE_PORT or 1521)")
    parser.add_argument("--service", default=None, help="Oracle service name (default: env ORACLE_SERVICE or FREEPDB1)")
    parser.add_argument("--user", default=None, help="Oracle user (default: env ORACLE_USER or benchmark)")
    parser.add_argument("--password", default=None, help="Oracle password (default: env ORACLE_PASSWORD or benchmark)")
    parser.add_argument("--rows", type=int, default=NUM_ROWS, help=f"Number of rows to insert (default: {NUM_ROWS})")
    parser.add_argument("--iterations", type=int, default=ITERATIONS, help=f"Query iterations for averaging (default: {ITERATIONS})")
    parser.add_argument("--no-cleanup", action="store_true", help="Keep benchmark tables after run")
    parser.add_argument("--no-fetch", action="store_true", help="Execute queries without fetching rows")
    return parser.parse_args()


def main():
    args = parse_args()

    global NUM_ROWS, ITERATIONS, FETCH_ALL
    NUM_ROWS = args.rows
    ITERATIONS = args.iterations
    FETCH_ALL = not args.no_fetch

    print_section("Oracle Benchmark Tool")
    print(f"  Rows: {NUM_ROWS}  |  Iterations: {ITERATIONS}  |  Fetch: {FETCH_ALL}")

    conn = get_connection(args)
    cursor = conn.cursor()
    # Increase array size for faster fetches
    cursor.arraysize = 5000

    try:
        # ---- Single Table Benchmark ----
        print_section("Phase 1: Single Table (100 columns)")
        print("\n[Setup]")
        create_single_table(cursor)
        conn.commit()

        print("\n[Insert Data]")
        with timer() as t_ins:
            insert_single_table(conn, cursor)
        print(f"  Insert time: {t_ins['elapsed']:.2f}s")

        print("\n[Benchmarking Reads]")
        single_results = benchmark_single_table(cursor)
        format_results(single_results)

        # ---- Multi-Table Benchmark ----
        print_section("Phase 2: Three Tables with JOIN (100 columns total)")
        print("\n[Setup]")
        create_multi_tables(cursor)
        conn.commit()

        print("\n[Insert Data]")
        with timer() as t_ins:
            insert_multi_tables(conn, cursor)
        print(f"  Insert time: {t_ins['elapsed']:.2f}s")

        print("\n[Benchmarking Reads with JOINs]")
        multi_results = benchmark_multi_tables(cursor)
        format_results(multi_results)

        # ---- Comparison ----
        print_section("Comparison: Single Table vs 3-Table JOIN")
        comparison = []
        for s, m in zip(single_results, multi_results):
            ratio = m["Avg (s)"] / s["Avg (s)"] if s["Avg (s)"] > 0 else float("inf")
            comparison.append({
                "Pattern": s["Pattern"].replace(" (5 cols)", "").replace(" *", " *"),
                "Single Avg (s)": s["Avg (s)"],
                "JOIN Avg (s)": m["Avg (s)"],
                "JOIN / Single Ratio": ratio,
                "Verdict": "JOIN slower" if ratio > 1.05 else ("Similar" if ratio > 0.95 else "JOIN faster"),
            })
        format_results(comparison)

    finally:
        if not args.no_cleanup:
            print_section("Cleanup")
            cleanup(cursor)
            conn.commit()
        cursor.close()
        conn.close()

    print_section("Done")


if __name__ == "__main__":
    main()

File: oracle-benchmark-tool-java/src/main/java/com/benchmark/OracleBenchmark.java
package com.benchmark;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.DoubleSummaryStatistics;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Oracle Read Performance Benchmark
 *
 * Compares read performance between:
 *   1) A single wide table with 100 columns -- always reads ALL columns (SELECT *)
 *   2) Three normalized tables (same data split across 3 tables) -- reads only
 *      NEEDED columns (10 cols) via a single JOIN of 2 tables.
 *
 * The point: with a wide table you must read everything, but with normalized
 * tables you only JOIN the tables that have the columns you need.
 */
public class OracleBenchmark {

    // Configuration
    private static final int NUM_COLUMNS = 100;
    private static final int NORM_SELECT_COLS = 10;
    private static int numRows = 10_000;
    private static int iterations = 10;
    private static boolean fetchAll = true;

    // Table split: 34 + 33 + 33 = 100
    private static final int[] SPLIT = {34, 33, 33};

    // Table/view names
    private static final String SINGLE_TABLE = "BENCH_SINGLE";
    private static final String SINGLE_VIEW = "BENCH_SINGLE_V";
    private static final String[] MULTI_TABLES = {"BENCH_MULTI_A", "BENCH_MULTI_B", "BENCH_MULTI_C"};
    private static final String MULTI_VIEW = "BENCH_MULTI_V";

    private static final Random RANDOM = new Random();
    private static final String CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------

    private static String randomString(int length) {
        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length; i++) {
            sb.append(CHARS.charAt(RANDOM.nextInt(CHARS.length())));
        }
        return sb.toString();
    }

    private static String colName(int index) {
        return String.format("COL_%03d", index);
    }

    private static void dropIfExists(Statement stmt, String objType, String name) {
        try {
            if ("TABLE".equals(objType)) {
                stmt.execute("DROP " + objType + " " + name + " PURGE");
            } else {
                stmt.execute("DROP " + objType + " " + name);
            }
        } catch (SQLException ignored) {
            // Object doesn't exist
        }
    }

    private static void printSection(String title) {
        System.out.println();
        System.out.println("=".repeat(70));
        System.out.println("  " + title);
        System.out.println("=".repeat(70));
    }

    /**
     * Returns (start, end) column index ranges for each of the 3 tables.
     */
    private static int[][] colRanges() {
        int s1 = 1, e1 = SPLIT[0];
        int s2 = e1 + 1, e2 = e1 + SPLIT[1];
        int s3 = e2 + 1, e3 = e2 + SPLIT[2];
        return new int[][]{{s1, e1}, {s2, e2}, {s3, e3}};
    }

    // ---------------------------------------------------------------------------
    // Result formatting
    // ---------------------------------------------------------------------------

    private static class BenchmarkResult {
        final String pattern;
        final double min;
        final double max;
        final double avg;
        final double median;
        final double stdev;

        BenchmarkResult(String pattern, List<Double> timings) {
            this.pattern = pattern;
            DoubleSummaryStatistics stats = timings.stream()
                    .mapToDouble(Double::doubleValue)
                    .summaryStatistics();
            this.min = stats.getMin();
            this.max = stats.getMax();
            this.avg = stats.getAverage();

            List<Double> sorted = timings.stream().sorted().collect(Collectors.toList());
            int n = sorted.size();
            this.median = (n % 2 == 0)
                    ? (sorted.get(n / 2 - 1) + sorted.get(n / 2)) / 2.0
                    : sorted.get(n / 2);

            double variance = timings.stream()
                    .mapToDouble(t -> (t - this.avg) * (t - this.avg))
                    .sum() / (n - 1);
            this.stdev = (n > 1) ? Math.sqrt(variance) : 0.0;
        }
    }

    private static void formatResults(List<BenchmarkResult> results) {
        // Calculate column widths
        int patternWidth = Math.max(7, results.stream()
                .mapToInt(r -> r.pattern.length())
                .max().orElse(7));

        String headerFmt = "| %-" + patternWidth + "s | %9s | %9s | %9s | %11s | %10s |%n";
        String rowFmt = "| %-" + patternWidth + "s | %9.4f | %9.4f | %9.4f | %11.4f | %10.4f |%n";
        String sep = "+" + "-".repeat(patternWidth + 2) + "+"
                + "-".repeat(11) + "+" + "-".repeat(11) + "+"
                + "-".repeat(11) + "+" + "-".repeat(13) + "+"
                + "-".repeat(12) + "+";

        System.out.println(sep);
        System.out.printf(headerFmt, "Pattern", "Min (s)", "Max (s)", "Avg (s)", "Median (s)", "Stdev (s)");
        System.out.println(sep);
        for (BenchmarkResult r : results) {
            System.out.printf(rowFmt, r.pattern, r.min, r.max, r.avg, r.median, r.stdev);
            System.out.println(sep);
        }
    }

    private static void formatComparison(List<BenchmarkResult> singleResults,
                                         List<BenchmarkResult> multiResults) {
        String sep = "+" + "-".repeat(27) + "+" + "-".repeat(18) + "+"
                + "-".repeat(27) + "+" + "-".repeat(22) + "+"
                + "-".repeat(21) + "+" + "-".repeat(21) + "+";

        System.out.println(sep);
        System.out.printf("| %-25s | %16s | %-25s | %20s | %19s | %-19s |%n",
                "Single-Table Pattern", "Single Avg (s)", "Normalized Pattern",
                "Normalized Avg (s)", "Norm / Wide Ratio", "Verdict");
        System.out.println(sep);

        for (int i = 0; i < singleResults.size() && i < multiResults.size(); i++) {
            BenchmarkResult s = singleResults.get(i);
            BenchmarkResult m = multiResults.get(i);
            double ratio = (s.avg > 0) ? m.avg / s.avg : Double.POSITIVE_INFINITY;
            String verdict;
            if (ratio > 1.05) {
                verdict = "Normalized slower";
            } else if (ratio > 0.95) {
                verdict = "Similar";
            } else {
                verdict = "Normalized faster";
            }
            System.out.printf("| %-25s | %16.4f | %-25s | %20.4f | %19.4f | %-19s |%n",
                    s.pattern, s.avg, m.pattern, m.avg, ratio, verdict);
            System.out.println(sep);
        }
    }

    // ---------------------------------------------------------------------------
    // Single wide table
    // ---------------------------------------------------------------------------

    private static void createSingleTable(Statement stmt) throws SQLException {
        dropIfExists(stmt, "VIEW", SINGLE_VIEW);
        dropIfExists(stmt, "TABLE", SINGLE_TABLE);

        StringBuilder cols = new StringBuilder();
        StringBuilder allCols = new StringBuilder();
        for (int i = 1; i <= NUM_COLUMNS; i++) {
            if (i > 1) {
                cols.append(", ");
                allCols.append(", ");
            }
            cols.append(colName(i)).append(" VARCHAR2(50)");
            allCols.append(colName(i));
        }

        stmt.execute("CREATE TABLE " + SINGLE_TABLE + " (ID NUMBER PRIMARY KEY, " + cols + ")");
        stmt.execute("CREATE VIEW " + SINGLE_VIEW + " AS SELECT ID, " + allCols + " FROM " + SINGLE_TABLE);

        System.out.println("  Created table  " + SINGLE_TABLE + " with " + NUM_COLUMNS + " columns");
        System.out.println("  Created view   " + SINGLE_VIEW + " (all " + NUM_COLUMNS + " columns)");
    }

    private static void insertSingleTable(Connection conn, Statement stmt) throws SQLException {
        StringBuilder colNames = new StringBuilder();
        StringBuilder placeholders = new StringBuilder();
        for (int i = 1; i <= NUM_COLUMNS; i++) {
            if (i > 1) {
                colNames.append(", ");
                placeholders.append(", ");
            }
            colNames.append(colName(i));
            placeholders.append("?");
        }

        String sql = "INSERT INTO " + SINGLE_TABLE + " (ID, " + colNames + ") VALUES (?, " + placeholders + ")";
        int batchSize = 1000;

        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            for (int start = 0; start < numRows; start += batchSize) {
                int end = Math.min(start + batchSize, numRows);
                for (int rowId = start + 1; rowId <= end; rowId++) {
                    ps.setInt(1, rowId);
                    for (int c = 1; c <= NUM_COLUMNS; c++) {
                        ps.setString(c + 1, randomString(20));
                    }
                    ps.addBatch();
                }
                ps.executeBatch();
            }
        }
        conn.commit();
        System.out.println("  Inserted " + numRows + " rows into " + SINGLE_TABLE);
    }

    private static List<BenchmarkResult> benchmarkSingleTable(Connection conn) throws SQLException {
        String allCols = IntStream.rangeClosed(1, NUM_COLUMNS)
                .mapToObj(OracleBenchmark::colName)
                .collect(Collectors.joining(", "));

        Map<String, String> queries = new LinkedHashMap<>();
        queries.put("SELECT (all " + NUM_COLUMNS + " cols)",
                "SELECT ID, " + allCols + " FROM " + SINGLE_TABLE);
        queries.put("VIEW   (all " + NUM_COLUMNS + " cols)",
                "SELECT * FROM " + SINGLE_VIEW);
        queries.put("SELECT *",
                "SELECT * FROM " + SINGLE_TABLE);

        return runBenchmark(conn, queries);
    }

    // ---------------------------------------------------------------------------
    // Normalized tables (3 tables, read via 1 JOIN of 2 tables)
    // ---------------------------------------------------------------------------

    private static void createMultiTables(Statement stmt) throws SQLException {
        dropIfExists(stmt, "VIEW", MULTI_VIEW);
        for (String tbl : MULTI_TABLES) {
            dropIfExists(stmt, "TABLE", tbl);
        }

        int[][] ranges = colRanges();
        for (int t = 0; t < MULTI_TABLES.length; t++) {
            int start = ranges[t][0];
            int end = ranges[t][1];
            StringBuilder cols = new StringBuilder();
            for (int i = start; i <= end; i++) {
                if (i > start) cols.append(", ");
                cols.append(colName(i)).append(" VARCHAR2(50)");
            }
            stmt.execute("CREATE TABLE " + MULTI_TABLES[t]
                    + " (ID NUMBER PRIMARY KEY, " + cols + ")");
            System.out.printf("  Created table %s with %d columns (COL_%03d..COL_%03d)%n",
                    MULTI_TABLES[t], end - start + 1, start, end);
        }

        // Create a view that JOINs only 2 tables (A and B)
        int[][] r = colRanges();
        int aStart = r[0][0];
        int bStart = r[1][0];
        int half = NORM_SELECT_COLS / 2;

        String viewACols = IntStream.range(aStart, aStart + half)
                .mapToObj(i -> "a." + colName(i))
                .collect(Collectors.joining(", "));
        String viewBCols = IntStream.range(bStart, bStart + (NORM_SELECT_COLS - half))
                .mapToObj(i -> "b." + colName(i))
                .collect(Collectors.joining(", "));

        stmt.execute("CREATE VIEW " + MULTI_VIEW + " AS SELECT a.ID, " + viewACols + ", " + viewBCols
                + " FROM " + MULTI_TABLES[0] + " a JOIN " + MULTI_TABLES[1] + " b ON a.ID = b.ID");
        System.out.println("  Created view  " + MULTI_VIEW + " (2-table JOIN, " + NORM_SELECT_COLS + " cols)");
    }

    private static void insertMultiTables(Connection conn) throws SQLException {
        int[][] ranges = colRanges();
        int batchSize = 1000;

        // Prepare statements for each table
        PreparedStatement[] psList = new PreparedStatement[MULTI_TABLES.length];
        for (int t = 0; t < MULTI_TABLES.length; t++) {
            int start = ranges[t][0];
            int end = ranges[t][1];
            int numCols = end - start + 1;
            StringBuilder colNames = new StringBuilder();
            StringBuilder placeholders = new StringBuilder();
            for (int i = start; i <= end; i++) {
                if (i > start) {
                    colNames.append(", ");
                    placeholders.append(", ");
                }
                colNames.append(colName(i));
                placeholders.append("?");
            }
            psList[t] = conn.prepareStatement(
                    "INSERT INTO " + MULTI_TABLES[t] + " (ID, " + colNames + ") VALUES (?, " + placeholders + ")");
        }

        for (int startRow = 0; startRow < numRows; startRow += batchSize) {
            int endRow = Math.min(startRow + batchSize, numRows);
            for (int rowId = startRow + 1; rowId <= endRow; rowId++) {
                // Generate all 100 column values for this row
                String[] allColValues = new String[NUM_COLUMNS];
                for (int c = 0; c < NUM_COLUMNS; c++) {
                    allColValues[c] = randomString(20);
                }

                for (int t = 0; t < MULTI_TABLES.length; t++) {
                    int start = ranges[t][0];
                    int end = ranges[t][1];
                    psList[t].setInt(1, rowId);
                    for (int i = start; i <= end; i++) {
                        psList[t].setString(i - start + 2, allColValues[i - 1]);
                    }
                    psList[t].addBatch();
                }
            }
            for (PreparedStatement ps : psList) {
                ps.executeBatch();
            }
        }
        conn.commit();

        for (PreparedStatement ps : psList) {
            ps.close();
        }
        System.out.println("  Inserted " + numRows + " rows into each of "
                + String.join(", ", MULTI_TABLES));
    }

    private static List<BenchmarkResult> benchmarkMultiTables(Connection conn) throws SQLException {
        int[][] ranges = colRanges();
        int aStart = ranges[0][0];
        int bStart = ranges[1][0];
        int half = NORM_SELECT_COLS / 2;

        String aCols = IntStream.range(aStart, aStart + half)
                .mapToObj(i -> "a." + colName(i))
                .collect(Collectors.joining(", "));
        String bCols = IntStream.range(bStart, bStart + (NORM_SELECT_COLS - half))
                .mapToObj(i -> "b." + colName(i))
                .collect(Collectors.joining(", "));
        String selectCols = aCols + ", " + bCols;

        String joinClause = "FROM " + MULTI_TABLES[0] + " a JOIN " + MULTI_TABLES[1] + " b ON a.ID = b.ID";

        Map<String, String> queries = new LinkedHashMap<>();
        queries.put("SELECT (" + NORM_SELECT_COLS + " cols) + JOIN",
                "SELECT a.ID, " + selectCols + " " + joinClause);
        queries.put("VIEW   (" + NORM_SELECT_COLS + " cols) + JOIN",
                "SELECT * FROM " + MULTI_VIEW);
        queries.put("SELECT * + JOIN",
                "SELECT * " + joinClause);

        return runBenchmark(conn, queries);
    }

    // ---------------------------------------------------------------------------
    // Common benchmark runner
    // ---------------------------------------------------------------------------

    private static List<BenchmarkResult> runBenchmark(Connection conn,
                                                      Map<String, String> queries) throws SQLException {
        List<BenchmarkResult> results = new ArrayList<>();

        for (Map.Entry<String, String> entry : queries.entrySet()) {
            String label = entry.getKey();
            String sql = entry.getValue();
            List<Double> timings = new ArrayList<>();

            for (int iter = 0; iter < iterations; iter++) {
                long startTime = System.nanoTime();
                try (Statement stmt = conn.createStatement()) {
                    stmt.setFetchSize(5000);
                    try (ResultSet rs = stmt.executeQuery(sql)) {
                        if (fetchAll) {
                            while (rs.next()) {
                                // consume all rows
                            }
                        }
                    }
                }
                double elapsed = (System.nanoTime() - startTime) / 1_000_000_000.0;
                timings.add(elapsed);
            }

            results.add(new BenchmarkResult(label, timings));
        }

        return results;
    }

    // ---------------------------------------------------------------------------
    // Cleanup
    // ---------------------------------------------------------------------------

    private static void cleanup(Statement stmt) {
        dropIfExists(stmt, "VIEW", SINGLE_VIEW);
        dropIfExists(stmt, "TABLE", SINGLE_TABLE);
        dropIfExists(stmt, "VIEW", MULTI_VIEW);
        for (String tbl : MULTI_TABLES) {
            dropIfExists(stmt, "TABLE", tbl);
        }
        System.out.println("  Cleaned up all benchmark objects.");
    }

    // ---------------------------------------------------------------------------
    // CLI argument parsing
    // ---------------------------------------------------------------------------

    private static Map<String, String> parseArgs(String[] args) {
        Map<String, String> params = new LinkedHashMap<>();
        params.put("host", envOrDefault("ORACLE_HOST", "localhost"));
        params.put("port", envOrDefault("ORACLE_PORT", "1521"));
        params.put("service", envOrDefault("ORACLE_SERVICE", "FREEPDB1"));
        params.put("user", envOrDefault("ORACLE_USER", "benchmark"));
        params.put("password", envOrDefault("ORACLE_PASSWORD", "benchmark"));
        params.put("rows", String.valueOf(numRows));
        params.put("iterations", String.valueOf(iterations));
        params.put("no-cleanup", "false");
        params.put("no-fetch", "false");

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--host" -> params.put("host", args[++i]);
                case "--port" -> params.put("port", args[++i]);
                case "--service" -> params.put("service", args[++i]);
                case "--user" -> params.put("user", args[++i]);
                case "--password" -> params.put("password", args[++i]);
                case "--rows" -> params.put("rows", args[++i]);
                case "--iterations" -> params.put("iterations", args[++i]);
                case "--no-cleanup" -> params.put("no-cleanup", "true");
                case "--no-fetch" -> params.put("no-fetch", "true");
                default -> {
                    System.err.println("Unknown argument: " + args[i]);
                    System.exit(1);
                }
            }
        }
        return params;
    }

    private static String envOrDefault(String envVar, String defaultVal) {
        String val = System.getenv(envVar);
        return (val != null && !val.isEmpty()) ? val : defaultVal;
    }

    // ---------------------------------------------------------------------------
    // Main
    // ---------------------------------------------------------------------------

    public static void main(String[] args) throws Exception {
        Map<String, String> params = parseArgs(args);

        numRows = Integer.parseInt(params.get("rows"));
        iterations = Integer.parseInt(params.get("iterations"));
        fetchAll = !"true".equals(params.get("no-fetch"));
        boolean noCleanup = "true".equals(params.get("no-cleanup"));

        String jdbcUrl = "jdbc:oracle:thin:@//" + params.get("host") + ":"
                + params.get("port") + "/" + params.get("service");

        printSection("Oracle Benchmark Tool (Java)");
        System.out.println("  Rows: " + numRows + "  |  Iterations: " + iterations + "  |  Fetch: " + fetchAll);
        System.out.println("  Wide table:       reads ALL " + NUM_COLUMNS + " columns");
        System.out.println("  Normalized:       reads only " + NORM_SELECT_COLS + " columns via 1 JOIN (2 tables)");

        try (Connection conn = DriverManager.getConnection(jdbcUrl, params.get("user"), params.get("password"))) {
            conn.setAutoCommit(false);
            Statement stmt = conn.createStatement();

            try {
                // ---- Single Table Benchmark ----
                printSection("Phase 1: Single Table -- read ALL " + NUM_COLUMNS + " columns");
                System.out.println("\n[Setup]");
                createSingleTable(stmt);
                conn.commit();

                System.out.println("\n[Insert Data]");
                long t0 = System.nanoTime();
                insertSingleTable(conn, stmt);
                double insertTime = (System.nanoTime() - t0) / 1_000_000_000.0;
                System.out.printf("  Insert time: %.2fs%n", insertTime);

                System.out.println("\n[Benchmarking Reads]");
                List<BenchmarkResult> singleResults = benchmarkSingleTable(conn);
                formatResults(singleResults);

                // ---- Multi-Table Benchmark ----
                printSection("Phase 2: Normalized Tables -- read only " + NORM_SELECT_COLS + " cols via 1 JOIN");
                System.out.println("\n[Setup]");
                createMultiTables(stmt);
                conn.commit();

                System.out.println("\n[Insert Data]");
                t0 = System.nanoTime();
                insertMultiTables(conn);
                insertTime = (System.nanoTime() - t0) / 1_000_000_000.0;
                System.out.printf("  Insert time: %.2fs%n", insertTime);

                System.out.println("\n[Benchmarking Reads (2-table JOIN, selected cols only)]");
                List<BenchmarkResult> multiResults = benchmarkMultiTables(conn);
                formatResults(multiResults);

                // ---- Comparison ----
                printSection("Comparison: Wide Table (all " + NUM_COLUMNS
                        + " cols) vs Normalized (" + NORM_SELECT_COLS + " cols)");
                formatComparison(singleResults, multiResults);

            } finally {
                if (!noCleanup) {
                    printSection("Cleanup");
                    cleanup(stmt);
                    conn.commit();
                }
                stmt.close();
            }
        }

        printSection("Done");
    }
}
